File: src/main/java/wd/EmployeesREST/service/EmployeeService.java
package wd.EmployeesREST.service;

import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.List;

import wd.EmployeesREST.exceptions.EmployeeServiceNotFoundException;
import wd.EmployeesREST.dao.EmployeeRepository;
import wd.EmployeesREST.dto.Employee;

@Service
public class EmployeeService {
    private final EmployeeRepository employeeRepository;

    public EmployeeService(EmployeeRepository employeeRepository) {
        this.employeeRepository = employeeRepository;
    }

    public Employee getEmployeeById(long employeeID) throws EmployeeServiceNotFoundException {
        Employee returnedEmployee = employeeRepository.findById(employeeID).orElseThrow(() ->
                new EmployeeServiceNotFoundException(String.format("Employee with id %d does not exist!", employeeID)));
        return returnedEmployee;
    }

    public List<Employee> getEmployeesByLastAndFirstName(String lastName, String firstName) throws EmployeeServiceNotFoundException {
        List<Employee> employeeSearchResult = employeeRepository.findByLastNameContainingIgnoreCaseAndFirstNameContainingIgnoreCase(lastName, firstName).stream().toList();
        if (employeeSearchResult.isEmpty()) {
            throw new EmployeeServiceNotFoundException(String.format("There is no employee with lastname %s and firstname %s", lastName, firstName));
        }
        return employeeSearchResult;
    }

    public void addEmployee(Employee newEmployee) {
        employeeRepository.save(newEmployee);
    }

    public void deleteEmployee(Long employeeToDeleteId) throws EmployeeServiceNotFoundException {
        employeeRepository.deleteById(employeeToDeleteId);
    }

    @Transactional
    public void update(Employee updatedEmployee, long employeeId) throws EmployeeServiceNotFoundException {
        employeeRepository.findById(employeeId).orElseThrow(() ->
                new EmployeeServiceNotFoundException(String.format("Employee with id %d does not exist!", employeeId)));

        employeeRepository.save(Employee.builder().employeeID(employeeId).firstName(updatedEmployee.getFirstName())
                .lastName(updatedEmployee.getLastName()).dateOfBirth(updatedEmployee.getDateOfBirth()).gender(updatedEmployee
                        .getGender()).departmentID(updatedEmployee.getDepartmentID()).jobTitle(updatedEmployee.getJobTitle()).build());
    }
}

File: src/main/kotlin/com/employees/service/EmployeeService.kt
package com.employees.service

import com.employees.dao.EmployeeRepository
import com.employees.dto.Employee
import com.employees.exceptions.EmployeeServiceNotFoundException
import com.employees.jms.EmployeeJmsProducer
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
open class EmployeeService(
    private val employeeRepository: EmployeeRepository,
    private val producer: EmployeeJmsProducer
) {

    @Throws(EmployeeServiceNotFoundException::class)
    fun getEmployeeById(employeeID: Long): Employee? {
        return employeeRepository.findById(employeeID).orElseThrow {
            EmployeeServiceNotFoundException(
                String.format(
                    "Employee with id %d does not exist!",
                    employeeID
                )
            )
        }
    }

    @Throws(EmployeeServiceNotFoundException::class)
    fun getEmployeesByLastAndFirstName(lastName: String?, firstName: String?): List<Employee?>? {
        return employeeRepository.findByLastNameContainingIgnoreCaseAndFirstNameContainingIgnoreCase(
            lastName,
            firstName
        )
    }

    fun addEmployee(newEmployee: Employee?) {
        newEmployee?.let { employeeRepository.save(it) }
    }

    @Throws(EmployeeServiceNotFoundException::class)
    fun deleteEmployee(employeeToDeleteId: Long?) {
        employeeRepository.deleteEmployeesByEmployeeID(employeeToDeleteId)
    }

    @Transactional
    @Throws(EmployeeServiceNotFoundException::class)
    open fun updateEmployee(updatedEmployee: Employee, employeeId: Long) {
        employeeRepository.findById(employeeId).orElseThrow {
            EmployeeServiceNotFoundException(
                String.format(
                    "Employee with id %d does not exist!",
                    employeeId
                )
            )
        }
        employeeRepository.save(
            Employee(
                employeeId,
                updatedEmployee.firstName,
                updatedEmployee.lastName,
                updatedEmployee.gender,
                updatedEmployee.dateOfBirth,
                updatedEmployee.jobTitle,
                updatedEmployee.departmentID
            )
        )
    }

    fun addEmployeeUsingJMS(employee: Employee?) {
        producer.addEmployee(employee)
    }
}